packages/mdc-menu-surface/foundation.js

```diff
@@ -0,0 +1,572 @@
+    const {bodyDimensions, windowScroll, viewport, viewportDistance} = this.measures_;
+
+    for (const prop in position) {
+      if (position.hasOwnProperty(prop)) {
+        // Hoisted surfaces need to have the anchor elements location on the page added to the
+        // position properties for proper alignment on the body.
+        if (viewportDistance.hasOwnProperty(prop)) {
+          position[prop] = parseInt(position[prop], 10) + viewportDistance[prop];
+        }
+
+        // Surfaces that are absolutely positioned need to have additional calculations for scroll
+        // and bottom positioning.
+        if (!this.isFixedPosition_ && prop === 'top') {
+          position[prop] = parseInt(position[prop], 10) + windowScroll.y;
+        } else if (!this.isFixedPosition_ && prop === 'bottom') {
+          position[prop] = bodyDimensions.height - (viewport.height + windowScroll.y) + parseInt(position[prop], 10);
+        }
+      }
+    }
+
+    return position;
+  }
+
+  /**
+   * Open the menu surface.
+   */
+  open() {
+    this.adapter_.saveFocus();
+
+    if (!this.quickOpen_) {
+      this.adapter_.addClass(MDCMenuSurfaceFoundation.cssClasses.ANIMATING_OPEN);
+    }
+
+    this.animationRequestId_ = requestAnimationFrame(() => {
+      this.adapter_.addClass(MDCMenuSurfaceFoundation.cssClasses.OPEN);
+      this.dimensions_ = this.adapter_.getInnerDimensions();
+      this.autoPosition_();
+      this.adapter_.registerBodyClickHandler(this.documentClickHandler_);
+      if (this.quickOpen_) {
+        this.adapter_.notifyOpen();
+      } else {
+        this.openAnimationEndTimerId_ = setTimeout(() => {
+          this.openAnimationEndTimerId_ = 0;
+          this.adapter_.removeClass(MDCMenuSurfaceFoundation.cssClasses.ANIMATING_OPEN);
+          this.adapter_.notifyOpen();
+        }, numbers.TRANSITION_OPEN_DURATION);
+      }
+    });
+    this.isOpen_ = true;
+  }
+
+  /**
+   * Closes the menu surface.
+   */
+  close() {
+    this.adapter_.deregisterBodyClickHandler(this.documentClickHandler_);
+
+    if (!this.quickOpen_) {
+      this.adapter_.addClass(MDCMenuSurfaceFoundation.cssClasses.ANIMATING_CLOSED);
+    }
+
+    requestAnimationFrame(() => {
+      this.adapter_.removeClass(MDCMenuSurfaceFoundation.cssClasses.OPEN);
+      if (this.quickOpen_) {
+        this.adapter_.notifyClose();
+      } else {
+        this.closeAnimationEndTimerId_ = setTimeout(() => {
+          this.closeAnimationEndTimerId_ = 0;
+          this.adapter_.removeClass(MDCMenuSurfaceFoundation.cssClasses.ANIMATING_CLOSED);
+          this.adapter_.notifyClose();
+        }, numbers.TRANSITION_CLOSE_DURATION);
+      }
+    });
+
+    this.isOpen_ = false;
+    this.shouldRestoreFocus_();
+  }
+
+  /**
+   * The last focused element when the menu surface was opened should regain focus, if the user is
+   * focused on or within the menu surface when it is closed.
+   * @private
+   */
+  shouldRestoreFocus_() {
```

Comment:
This is a weird name for how this method works. I'd expect this to return a boolean and for `close` to perform the actual adapter call. Otherwise rename this to something like `maybeRestoreFocus_` at least.